Sources/CompositeData/CompositeDataSource.swift

```diff
@@ -0,0 +1,118 @@
+import Foundation
+
+public class CompositeDataSource<A: DataSourceType, B: DataSourceType>: DataSourceType
+{
+    public var updateHandler = IndexedUpdateHandler()
+
+    let firstDataSource: A
+    let secondDataSource: B
+
+    private var firstDataSourceSubscription: IndexedUpdateHandler.Subscription?
+    private var secondDataSourceSubscription: IndexedUpdateHandler.Subscription?
+
+    public init(firstDataSource: A, secondDataSource: B) {
+        self.firstDataSource = firstDataSource
+        self.secondDataSource = secondDataSource
+
+        firstDataSourceSubscription = firstDataSource.updateHandler.subscribe { [weak self] update in
+            guard let self = self else { return }
+            print("firstDataSourceSubscription update: \(update)")
+            self.updateHandler.send(update: update)
+        }
+
+        secondDataSourceSubscription = secondDataSource.updateHandler.subscribe { [weak self] update in
+            guard let self = self else { return }
+            print("secondDataSourceSubscription update: \(update)")
+            let mappedUpdate = update.offsetSections(by: firstDataSource.numberOfSections())
+            print("secondDataSourceSubscription update: \(mappedUpdate)")
+            self.updateHandler.send(update: mappedUpdate)
+        }
+    }
+}
+
+public extension CompositeDataSource
+{
+    enum Item {
+        case A(A.Item)
+        case B(B.Item)
```

Comment:
The call site ergonomics are giving me pause. As a simple example, imagine we have an `Int` data source for "amounts", and a `String` data source for "names".. If we have to query the data source, then it's something like:

```
switch dataSource.item(at: indexPath) {
  case let .A(amount): …
  case let .B(name): …
}
```

The `A`/`B` really stick out for me and don't convey much meaning. Making a bespoke datasource generic over a custom `enum` can result in something like…

```
switch dataSource.item(at: indexPath) {
  case let .amount(amount): …
  case let .name(name): …
}
```

This much more semantic type information and is more understandable imo. I don't have a great idea on how to get closer to this though, just a thought/concern. @heiberg any thoughts on this?

Reply:
I agree with the pause, and wanted to bring in eyes before trying to perfect it.

Maybe changing the naming of the casing to `case first(A.Item); case second(B.Item)` is an improvement, or maybe it's more confusing 🤷🏻‍♂️  
At least the language would be consistent with the constructor with that change though 🤔

Reply:
I'm not really sold on the forced `Item` type here. But I added a separate comment with a possible approach we could take for this.